[jit] allow slicing multiple dimensions with indices

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45239 [WIP] [jit] allow slicing multiple dimensions with indices**

Differential Revision: [D23886919](https://our.internmc.facebook.com/intern/diff/D23886919)

Commit:
[jit] allow slicing multiple dimensions with indicies

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_list_dict.py:
```diff
@@ -408,6 +408,34 @@ def test_over_slice():
+            x = x[[0, 1], :, [1]]
+            return x
+
+        test_func(test_index_slice1, (a,))
```
You can use self.checkScript for same functionality.

Lilyjjo replied:
Ah, I didn't use that because I wasn't sure if it was comparing the outputs of the scripted vs non-scripted modules. I double checked on the self.checkScript code and I see that it actually does. I'll change it to use that instead :)

gmagogsfm commented on test/jit/test_list_dict.py:
```diff
@@ -408,6 +408,34 @@ def test_over_slice():
+            ]
+        )
+
+        def test_index_slice1(x):
```
Do we have proper behaviors defined for corner cases like 1) empty list 2) out of bound indices etc?

Commit:
Update on "[jit] allow slicing multiple dimensions with indicies"


Differential Revision: [D23886919](https://our.internmc.facebook.com/intern/diff/D23886919)

[ghstack-poisoned]

Commit:
Update on "[WIP] [jit] allow slicing multiple dimensions with indices "


Currently investigating issue where indexing with an empty list in eager returns a tensor with a zero-sized dim in the empty list spot *OTOH* the scripted module will throw this error:
" RuntimeError: The following operation failed in the TorchScript interpreter.
    Traceback of TorchScript (most recent call last):
    RuntimeError: Input must be of ints, floats, or bools, got Tensor
    Empty lists default to List[Tensor]. Add a variable annotation to the
    assignment to create an empty list of another type
    (torch.jit.annotate(List[T, []]) where T is the type of elements in the
    list for Python 2) "

Differential Revision: [D23886919](https://our.internmc.facebook.com/intern/diff/D23886919)

[ghstack-poisoned]

Commit:
Update on "[WIP] [jit] allow slicing multiple dimensions with indices "


Currently investigating issue where indexing with an empty list in eager returns a tensor with a zero-sized dim in the empty list spot *OTOH* the scripted module will throw this error:
" RuntimeError: The following operation failed in the TorchScript interpreter.
    Traceback of TorchScript (most recent call last):
    RuntimeError: Input must be of ints, floats, or bools, got Tensor
    Empty lists default to List[Tensor]. Add a variable annotation to the
    assignment to create an empty list of another type
    (torch.jit.annotate(List[T, []]) where T is the type of elements in the
    list for Python 2) "

Differential Revision: [D23886919](https://our.internmc.facebook.com/intern/diff/D23886919)

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_list_dict.py:
```diff
@@ -408,6 +408,49 @@ def test_over_slice():
+            x = x[[], :, :]
+            return x
+
+        self.checkScript(test_index_slice1, (a,))
```
Move check right next to its function under test, ideally all of them should become individual small testcases

gmagogsfm commented on test/jit/test_list_dict.py:
```diff
@@ -408,6 +408,49 @@ def test_over_slice():
+        with self.assertRaisesRegex(RuntimeError, "index 4 is out of bounds for dimension 0 with size 3"):
+            self.checkScript(test_index_slice4, (a,))
+
+        with self.assertRaises(RuntimeError):
```
I don't think this test case is needed, please make sure empty sequence with torch.jit.annotate works.

Commit:
Update on "[WIP] [jit] allow slicing multiple dimensions with indices "


Differential Revision: [D23886919](https://our.internmc.facebook.com/intern/diff/D23886919)

[ghstack-poisoned]
